File: day-2/main.py
def load_data(file):
    data = [[*map(int, line.strip().split())] for line in open(file)]

    return data


def level_safe(level: list):

    new_list = level.copy()
    current_level = new_list.pop(0)

    if current_level < new_list[0]:
        increasing = True
    else:
        increasing = False

    while len(new_list) != 0:

        difference = 0
        if increasing:
            if current_level > new_list[0]:
                return False
            difference = new_list[0] - current_level
        else:
            if current_level < new_list[0]:
                return False
            difference = current_level - new_list[0]

        if abs(difference) == 0 or abs(difference) > 3:
            return False

        current_level = new_list.pop(0)

    return True


def main():
    # data = load_data("input_test.txt")
    data = load_data("input.txt")
    safe_levels = 0

    for l in data:

        if level_safe(l):
            safe_levels += 1
        # for part two
        # else:
        #     for i in range(len(l)):
        #         temp = l.copy()
        #         temp.pop(i)
        #         if level_safe(temp):
        #             safe_levels += 1
        #             break
    print(safe_levels)


main()

File: day-2/main.go
package main

import (
	"fmt"
	"math"
	"strconv"
	"strings"

	"git.thecodedom.com/aoc2024/tools"
)

type intSlice struct {
	number []int
}

func createSlices(data []byte) []intSlice {

	test := string(data)
	split := strings.Split(test, "\n")

	var numbers []intSlice

	for _, num := range split {
		if num != "" {

			var ints []int
			s := strings.Split(num, " ")
			for _, n := range s {
				t, err := strconv.Atoi(n)
				if err != nil {
					panic(err)
				}
				ints = append(ints, t)
			}
			numbers = append(numbers, intSlice{number: ints})

		}

	}

	return numbers
}

func (i intSlice) levelSafe() bool {
	var increasing bool = true

	currentLevel, temp := i.number[0], i.number[1:]
	i.number = temp

	if currentLevel > i.number[0] {
		increasing = false
	}

	for len(i.number) > 0 {
		difference := 0

		if increasing == true {
			if currentLevel > i.number[0] {
				return false
			}
			difference = currentLevel - i.number[0]
		} else {
			if currentLevel < i.number[0] {
				return false
			}
			difference = i.number[0] - currentLevel
		}

		if math.Abs(float64(difference)) == 0 || math.Abs(float64(difference)) > 3 {
			return false
		}

		currentLevel, temp = i.number[0], i.number[1:]
		i.number = temp

	}
	return true

}

func main() {
	d := tools.LoadData("input.txt")
	s := createSlices(d)

	var totalSafe int = 0

	for _, o := range s {
		r := o.levelSafe()

		if r {
			totalSafe += 1
		}
	}

	fmt.Println(totalSafe)
}
